### PR TITLE
[SPARK-47719][SQL] Change spark.sql.legacy.timeParserPolicy default to CORRECTED

### PR DESCRIPTION
**Pull request by srielau**
### What changes were proposed in this pull request?

We changed the time parser policy in Spark 3.0.0.
The config has since defaulted to raise an exception if there is a potential conflict between teh legacy and the new policy.
Spark 4.0.0 is a good time to default to the new policy

### Why are the changes needed?

Move the product forward and retire legacy behavior over time.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Run existing unit tests and verify changes.

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 1]
```diff
@@ -4016,13 +4016,13 @@ object SQLConf {
...
-      "The default value is EXCEPTION, RuntimeException is thrown when we will get different " +
-      "results.")
+      "When set to EXCEPTION, RuntimeException is thrown when we will get different " +
+      "results. The default is CORRECTED.")
     .version("3.0.0")
     .stringConf
     .transform(_.toUpperCase(Locale.ROOT))
     .checkValues(LegacyBehaviorPolicy.values.map(_.toString))
-    .createWithDefault(LegacyBehaviorPolicy.EXCEPTION.toString)
+    .createWithDefault(LegacyBehaviorPolicy.CORRECTED.toString)
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:4025]: Would have to update SQL migration guide at https://github.com/apache/spark/blob/master/docs/sql-migration-guide.md

**Comment by HyukjinKwon** [comment]
and this seems a duplicate of https://github.com/apache/spark/pull/45853 cc @cloud-fan

File: sql/core/src/test/resources/sql-tests/inputs/datetime-formatting-exception.sql [hunk 2]
```diff
@@ -0,0 +1,2 @@
+--SET spark.sql.legacy.timeParserPolicy=EXCEPTION
+--IMPORT datetime-formatting.sql
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/datetime-formatting-exception.sql:2]: I don't see any exception in the result golden file. Do we really need to test it again with EXCEPTION policy?

File: docs/sql-migration-guide.md [hunk 3]
```diff
@@ -46,6 +46,8 @@ license: |
 - Since Spark 4.0, MySQL JDBC datasource will read FLOAT as FloatType, while in Spark 3.5 and previous, it was read as DoubleType. To restore the previous behavior, you can cast the column to the old type.
 - Since Spark 4.0, MySQL JDBC datasource will read BIT(n > 1) as BinaryType, while in Spark 3.5 and previous, read as LongType. To restore the previous behavior, set `spark.sql.legacy.mysql.bitArrayMapping.enabled` to `true`.
 - Since Spark 4.0, MySQL JDBC datasource will write ShortType as SMALLINT, while in Spark 3.5 and previous, write as INTEGER. To restore the previous behavior, you can replace the column with IntegerType whenever before writing.
+- Since Spark 4.0, The default value for `spark.sql.legacy.ctePrecedencePolicy` has been changed from `EXCEPTION` to `CORRECTED`. Instead of raising an error, inner CTE definitions take precedence over outer definitions.
```
Comment (cloud-fan) [inline docs/sql-migration-guide.md:49]: to confirm: this change is already made and this PR is just filling the missing migration guide?

Comment (srielau) [reply docs/sql-migration-guide.md:49]: Correct

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 1 above)
Comment (srielau) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:4025]: Done

File: sql/core/src/test/resources/sql-tests/inputs/datetime-formatting-exception.sql (hunk 2 above)
Comment (srielau) [reply sql/core/src/test/resources/sql-tests/inputs/datetime-formatting-exception.sql:2]: I'm puzzling about this as well.
I cannot find where we explicitly tested CORRECTED before, although EXCEPTION was the default.
I expected copy the existing tests, label them EXCEPTION and it should behave as before.
And the existing files, with the new default should have no/fewer exceptions. This is not what I observe...

Comment (srielau) [reply sql/core/src/test/resources/sql-tests/inputs/datetime-formatting-exception.sql]: OK, I have a better understanding of what CORRECTED actually does.
Removed all the extra files.
I also found that there was SQ:ApiConf.scala needed to be fixed. That flushed out the expected changes for CORRECTED

**Comment by cloud-fan** [comment]
```
======================================================================
FAIL [2.071s]: test_error_enrichment_jvm_stacktrace (pyspark.sql.tests.connect.test_connect_session.SparkConnectSessionTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/__w/spark/spark/python/pyspark/sql/tests/connect/test_connect_session.py", line 137, in test_error_enrichment_jvm_stacktrace
    self.spark.sql(
AssertionError: SparkUpgradeException not raised
```


We need to update one pyspark test as well.

**Comment by srielau** [comment]
@gengliangwang @cloud-fan I think it's ready now. (one spark connect agrees test suite).

**Review by gengliangwang — APPROVED** [review]

**Comment by gengliangwang** [comment]
Thanks, merging to master. 
cc @dongjoon-hyun @yaooqinn who approved https://github.com/apache/spark/pull/45853

**Pull request closed without merging**